File: data_structures/src/KeyPatterns/Hashes/HashMap/java/WordPattern.java
package KeyPatterns.Hashes.HashMap.java;

import java.util.HashMap;
import java.util.Map;

/**
 * Word Pattern
 * - Given a pattern string `pattern` and a string `s`, determine if `s` follows
 * the same pattern.
 * - Each letter in `pattern` maps to exactly one unique word in `s`.
 * - Each unique word in `s` maps to exactly one letter in `pattern`.
 * 
 * ## Example
 * - Input: pattern = "abba", s = "dog cat cat dog"
 * - 'a' -> "dog"
 * - 'b' -> "cat"
 * - 'b' -> "cat"
 * - 'a' -> "dog" // consistent
 */
public class WordPattern {
    public static void main(String[] args) {
        System.out.println(wordPattern("abba", "dog cat cat dog"));
    }

    private static boolean wordPattern(String pattern, String s) {
        String[] words = s.split(" ");
        if (pattern.length() != words.length) {
            return false;
        }

        Map<Character, String> charToWord = new HashMap<>();
        Map<String, Character> wordToChar = new HashMap<>();

        for (int i = 0; i < pattern.length(); i++) {
            char ch = pattern.charAt(i);
            String word = words[i];

            if (charToWord.containsKey(ch) && !charToWord.get(ch).equals(word)) {
                return false;
            }
            if (wordToChar.containsKey(word) && wordToChar.get(word) != ch) {
                return false;
            }

            charToWord.put(ch, word);
            wordToChar.put(word, ch);
        }

        return true;
    }
}

File: data_structures/src/KeyPatterns/Hashes/HashMap/typescript/word_pattern.ts
function wordPattern(pattern: string, s: string) {
  const words = s.split(' ');
  if (pattern.length !== words.length) return false;

  const charToWord = new Map<string, string>();
  const wordToChar = new Map<string, string>();

  for (let i = 0; i < pattern.length; i++) {
    const ch = pattern.charAt(i);
    const word = words[i];

    if (charToWord.has(ch) && charToWord.get(ch) !== word) {
      return false;
    }

    if (wordToChar.has(word) && wordToChar.get(word) !== ch) {
      return false;
    }

    charToWord.set(ch, word);
    wordToChar.set(word, ch);
  }

  return true;
}
console.log(wordPattern('abba', 'dog cat cat dog'));
